core/control-plane/lib/control-plane-policies-lib/src/main/java/org/eclipse/edc/connector/controlplane/policy/contract/ContractExpiryCheck.java

```diff
@@ -62,48 +60,38 @@
  * </ul>
  * Please note that all {@link Operator}s except {@link Operator#IN} are supported.
  */
-public class ContractExpiryCheckFunction implements AtomicConstraintFunction<Permission> {
+public class ContractExpiryCheck {
 
     public static final String CONTRACT_EXPIRY_EVALUATION_KEY = EDC_NAMESPACE + "inForceDate";
     private static final String EXPRESSION_REGEX = "(contract[A,a]greement)\\+(-?[0-9]+)(s|m|h|d)";
     private static final int REGEX_GROUP_NUMERIC = 2;
     private static final int REGEX_GROUP_UNIT = 3;
 
-
-    @Override
-    public boolean evaluate(Operator operator, Object rightValue, Permission rule, PolicyContext context) {
-        if (!(rightValue instanceof String)) {
-            context.reportProblem("Right-value expected to be String but was " + rightValue.getClass());
-            return false;
+    public Result<Void> evaluate(Operator operator, Object rightValue, Instant now, ContractAgreement agreement) {
+        if (rightValue == null) {
+            return Result.failure("Right-value is null.");
         }
 
-        try {
-            var now = getContextData(Instant.class, context);
-
-            var rightValueStr = (String) rightValue;
-            var bound = asInstant(rightValueStr);
-            if (bound != null) {
-                return checkFixedPeriod(now, operator, bound);
-            }
-
-            var duration = asDuration(rightValueStr);
-            if (duration != null) {
-                var agreement = getContextData(ContractAgreement.class, context);
-                var signingDate = Instant.ofEpochSecond(agreement.getContractSigningDate());
-                return checkFixedPeriod(now, operator, signingDate.plus(duration));
-            }
-
-            context.reportProblem(format("Unsupported right-value, expected either an ISO-8061 String or a expression matching '%s', but got '%s'",
-                    CONTRACT_EXPIRY_EVALUATION_KEY, rightValueStr));
-
-        } catch (NullPointerException | DateTimeParseException ex) {
-            context.reportProblem(ex.getMessage());
+        if (!(rightValue instanceof String rightValueStr)) {
+            return Result.failure("Right-value expected to be String but was " + rightValue.getClass());
         }
-        return false;
+
+        return Optional.ofNullable(asInstant(rightValueStr))
+                .or(() -> Optional.ofNullable(asDuration(rightValueStr))
+                        .map(duration -> Instant.ofEpochSecond(agreement.getContractSigningDate())
+                                .plus(duration)
+                        )
+                ).map(bound -> checkFixedPeriod(now, operator, bound))
+                .map(it -> it ? Result.success() : Result.<Void>failure(""))
```

Comment:
To me, returning a `Result<Boolean>` from this method would feel more natural. There is a semantic difference between the check failing "structurally" (e.g. because malformed right-operand), and failing "semantically" (contract expired). WDYT?

Anyway, there should be some information on the failure:
```suggestion
                .map(it -> it ? Result.success() : Result.<Void>failure("Contract expired!"))
```

Reply:
yes, in fact I think I modeled this in an overcomplicated way (I was on a rush focusing to have everything working :sweat_smile: ), now I re-refactored in a single "function" class and it works in the same way as before (in case of expiration `false` is returned without messages), but maybe a message would be appreciated (e.g. to be appended to the transfer error detail).